Implement MethodHandleNatives native code

fengxue-IS opened the pull request:
Related: #7352 

Co-authored-by: Babneet Singh <sbabneet@ca.ibm.com>
Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
I've noticed that `Java_java_lang_invoke_MethodHandleNatives_checkClassBytes` have been added as part of Java 15, will have to look into if that should be merged into MethodHandleNatives.cpp

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+#define MN_REFERENCE_KIND_MASK	0xF
+
+#define MN_SEARCH_SUPERCLASSES	0x00100000
+#define MN_SEARCH_INTERFACES	0x00200000
```
What are these constants based on?  Do you match corresponding values in the javacode?  If so, please point add a comment referencing that

fengxue-IS replied:
added reference to Java constant

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(fieldID->field);
+
+		nameObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(name), (U_32)J9UTF8_LENGTH(name), J9_STR_INTERN);
+		typeObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(signature), (U_32)J9UTF8_LENGTH(signature), J9_STR_INTERN);
```
`nameObject` is corrupted if a GC occurs during the second object allocation.  The object pointer needs to be pushed in a special frame or stored somewhere the GC can find it and refreshed after the second allocate

fengxue-IS replied:
fixed

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+
+	if (result) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		char* signature = (char*)j9mem_allocate_memory(2 * sizeof(char*), OMRMEM_CATEGORY_VM);
```
```suggestion
		char* signature = (char*)j9mem_allocate_memory(2, OMRMEM_CATEGORY_VM);
```
The current code is allocating two pointers worth of data (ie: 2*8 bytes on a 64bit system)

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+	U_32 len = J9UTF8_LENGTH(romName);
+	char * name =  (char *)J9UTF8_DATA(romName);
+	U_32 length = len + numDims;
+	if (* name != '[')
+		length += 2;
```
There's a lot of missing `{` & `}`.  Please always use them for control flow as it prevents errors when modifying the code.
```suggestion
	if (* name != '[') {
		length += 2;
	}
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+	// Copy return type signature to descriptor string
+	strcpy(cursor, rSignature);
+	j9mem_free_memory(signatures);
+	return methodDescriptor;
```
This is really allocation heavy code.  Can the sizes be calculated up front and then the signatures written into a single allocation?

fengxue-IS replied:
Calculating the size would require two passes though the MethodType arguments list, not sure if that is better approach with all the heap object access.
I will add this to the todo list for performance investigation

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+		return NULL;
+	}
+
+	initStatus = jlrFieldClass->initializeStatus;
+	if (initStatus != J9ClassInitSucceeded && initStatus != (UDATA) vmThread) {
```
```suggestion
	if (VM_VMHelpers::classRequiresInitialization(vmThread, jlrFieldClass)) {
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+	J9Class *typeClass = NULL;
+	j9object_t fieldObject = NULL;
+	J9Class *jlrFieldClass = J9VMJAVALANGREFLECTFIELD(vmThread->javaVM);
+	UDATA initStatus;
```
```suggestion
	UDATA initStatus = 0;
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1434 @@
+}
+
+static j9object_t
+createFieldFromID(struct J9VMThread *vmThread, J9JNIFieldID *j9FieldID)
```
was this duplicated from the reflectHelpers code?  Any reason to make another copy of it?

fengxue-IS replied:
I wasn't sure if I should expand the reflection function table to export this API as this will be the single place to use it outside of reflecthelp.c, so I made the copy.

DanHeidinga replied:
Better to expand the table than to have to maintain a duplicate of the code.  Otherwise, it's easy to miss updating one of the two places if we find a bug

DanHeidinga reviewed (commented):
I reviewed about 1/2 this before stopping as I need more context on what the operations are expected to do.  

A good next step would be to document each of the functions so it's clear what the expected behaviour is

fengxue-IS commented:
Thanks for the review @DanHeidinga, I will add documentation comments for the function to provide details on the expected behavior

fengxue-IS reviewed (commented):
I have updated the code to include a function header that describe the expected behavior of each function and added comments to large code block to make it more readable. Also fixed some formatting issues.

While I was updating the comments, I realized there are more potential places where GC could occur and invalidate the native object reference. I will push the fix for these (either refetch the data or push objects to special frame) once local testing passed

DanHeidinga commented on runtime/jcl/uma/java_lang_invoke_MethodHandleNatives_exports.xml:
```diff
@@ -0,0 +1,36 @@
+	SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+<exports group="java_lang_invoke_MethodHandleNatives">
+		<export name="Java_java_lang_invoke_MethodHandleNatives_init" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_expand" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_resolve" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_getMembers" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_objectFieldOffset" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_staticFieldOffset" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_staticFieldBase" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_getMemberVMInfo" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetNormal" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetVolatile" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_copyOutBootstrapArguments" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_clearCallSiteContext" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_getNamedCon" />
+		<export name="Java_java_lang_invoke_MethodHandleNatives_registerNatives" />
```
Nitpick: overly indented inner code
```suggestion
	<export name="Java_java_lang_invoke_MethodHandleNatives_init" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_expand" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_resolve" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_getMembers" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_objectFieldOffset" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_staticFieldOffset" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_staticFieldBase" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_getMemberVMInfo" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetNormal" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetVolatile" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_copyOutBootstrapArguments" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_clearCallSiteContext" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_getNamedCon" />
	<export name="Java_java_lang_invoke_MethodHandleNatives_registerNatives" />
```

DanHeidinga commented on runtime/oti/jclprots.h:
```diff
@@ -943,6 +943,22 @@ extern J9_CFUNC void JNICALL
 Java_java_lang_invoke_MethodHandleNatives_checkClassBytes(JNIEnv *env, jclass jlClass, jbyteArray classRep);
 #endif /* JAVA_SPEC_VERSION >= 15 */
 
+/* java_lang_invoke_MethodHandleNatives.cpp */
```
Shouldn't these be `#if define(....OPENJDK_METHODHANDLES)` ?

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+jint JNICALL
+Java_java_lang_invoke_MethodHandleNatives_getNamedCon(JNIEnv *env, jclass clazz, jint which, jobjectArray name)
+{
+	return 0;
```
This needs an implementation or we'll fail to start when java assertions (`-ea`) are enabled.  @fengxue-IS can you create a issue to track implementing this?

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+void JNICALL
+Java_java_lang_invoke_MethodHandleNatives_registerNatives(JNIEnv *env, jclass clazz)
+{
+	return;
```
Why export the functions and make all the prototype changes to different headers rather than using the `registerNatives` function?  Given we're going to have to call it, using it would seem worthwhile....

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+
+extern "C" {
+
+/* Constants mapped from java.lang.invoke.MethodHandleNatives$Constants */
```
```suggestion
/* Constants mapped from java.lang.invoke.MethodHandleNatives$Constants 
 * These constants are validated by the MethodHandleNatives$Constants.verifyConstants()
 * method when Java assertions are enabled
 */
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+#define MN_CALLER_SENSITIVE	0x00100000
+
+#define MN_REFERENCE_KIND_SHIFT	24
+#define MN_REFERENCE_KIND_MASK	0xF
```
It would be good to indicate that the mask is shifted by the `MN_REFERENCE_KIND_SHIFT`

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+
+	Trc_JCL_java_lang_invoke_MethodHandleNatives_init_Entry(env, self, ref);
+	if (NULL == self || NULL == ref) {
```
Formatting
```suggestion
	if ((NULL == self) || (NULL == ref)) {
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+void
+initImpl(J9VMThread *currentThread, j9object_t membernameObject, j9object_t refObject)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
```
It's a good habit to mark these pointers as `const`
```suggestion
	const J9JavaVM *vm = currentThread->javaVM;
	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+ *
+ * Then for both, compute the MN.flags using access flags and invocation type based on the JNI-id.
+ *
+ * Throw an IllegalArgumentException if the refObject is not a Field/Method/Constructor
```
This should state that the caller must have vmaccess

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+		clazzObject = J9VM_J9CLASS_TO_HEAPCLASS(fieldID->declaringClass);
+
+		J9VMJAVALANGINVOKEMEMBERNAME_SET_NAME(currentThread, membernameObject, nameObject);
+		J9VMJAVALANGINVOKEMEMBERNAME_SET_TYPE(currentThread, membernameObject, typeObject);
```
Name and Type are only set here for field handles.  Is this deliberate?  Shouldn't methods and constructors also have a name and (method)type?

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+		J9UTF8 *name = J9ROMFIELDSHAPE_NAME(fieldID->field);
+		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(fieldID->field);
+
+		nameObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(name), (U_32)J9UTF8_LENGTH(name), J9_STR_INTERN);
```
A reflect field already has a name.  Can you reuse it here rather than allocate another copy of it?
```suggestion
		nameObject = J9VMJAVALANGREFLECTFIELD_NAME(vmThread, refObject);
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+		nameObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(name), (U_32)J9UTF8_LENGTH(name), J9_STR_INTERN);
+
+		PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, nameObject);
+		typeObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(signature), (U_32)J9UTF8_LENGTH(signature), J9_STR_INTERN);
```
Does `typeObject` need to be a string or can you set the class directly?  ie:
```suggestion
		typeObject =J9VMJAVALANGREFLECTFIELD_TYPE(vmThread, refObject);
```

DanHeidinga reviewed (commented):
Taking a break for lunch.  Will pick up the review from `sigForPrimitiveOrVoid` later

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+	return NULL;
+}
+
+char *
```
```suggestion
static char *
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+		result = 'V';
+	}
+
+	if (result) {
```
```suggestion
	if (result != 0) {
```

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+
+	if (result) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		char* signature = (char*)j9mem_allocate_memory(2, OMRMEM_CATEGORY_VM);
```
```suggestion
		char* signature = (char*)j9mem_allocate_memory(2, OMRMEM_CATEGORY_VM);
		if (signature == NULL) {
			// do something - set native oom and return?
		}
```

DanHeidinga replied:
If this is setting oom directly, then callers need to be protecting their objects.  Maybe better to return an error code for oom and check in the callers so it can be propagated back to the correct spot

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+	for (U_32 i = 0; i < numArgs; i++) {
+		j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
+		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
+		signatures[i] = sigForPrimitiveOrVoid(vm, pclass);
+		if (!signatures[i]) {
```
Needs to handle the NativeOOM case as well - likely by checking for a pending exception

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+	}
+
+	length++; //for null-termination
+	char * sig = (char *)j9mem_allocate_memory(length, OMRMEM_CATEGORY_VM);
```
malloc with no check for returning null

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+	j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+	J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+	char* rSignature = sigForPrimitiveOrVoid(vm, rclass);
+	if (!rSignature) {
```
```suggestion
	if (rSignature == NULL) {
```

DanHeidinga replied:
rSignature isn't a boolean type so `!rSignature` isn't an appropriate check

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+
+	*signatureLength += strlen(rSignature);
+
+	char* methodDescriptor = (char*)j9mem_allocate_memory(*signatureLength+1, OMRMEM_CATEGORY_VM);
```
missing null check

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+	J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+	char* rSignature = sigForPrimitiveOrVoid(vm, rclass);
+	if (!rSignature) {
+		rSignature = getClassSignature(currentThread, rclass);
```
This needs to communicate allocation failures out as well

DanHeidinga reviewed (commented):
@fengxue-IS there's a lot of feedback here already so I'll stop reviewing at this point.

Please look at all the allocations and check for allocations failures.  Figuring out how to report that back to the callers will help clarify your apis.

It's also good to minimize the lifetimes of every fetched object so they are less likely to be lost by a GC.  It'll also make the code easier to reason about

fengxue-IS commented:
Thanks @DanHeidinga for the indepth review, I will look into the issues pointed out and also go through all the allocation codes to make sure to check for any potential failures.

fengxue-IS commented:
- Added checks for allocation failure and handle exception from helper functions
- Added support for CallerSensitive annotation
- Moved `createFieldObject` to reflecthelp.c and added to the `J9ReflectionFunctions` table
- Modified variable declaration to minimize refetch required after potential GC points
- Use existing objects from reflect object rather than generating new objects
- Fixed formating
- Created #11443 to track todo items to be completed in future PR

@DanHeidinga can you take another look please

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1531 @@
+
+	j9object_t result = NULL;
+
+	J9RAMSingleSlotConstantRef *ramCP = (J9RAMSingleSlotConstantRef*)ramConstantPool + cpIndex;
```
It is better to get constants {int, float, double, long} from the ROM cp as they may be in the RAM cp.  When possible, we put constant data to the end of the ROM CP and have smaller RAM CPs.  Always fetch that kind of data from the ROM cp entries - see `getSingleSlotConstant` or `getDouleSlotConstant` from sun_reflect_constantpool.c

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1531 @@
+												methodObj = vm->reflectFunctions.createMethodObject(currentMethod, currentClass, NULL, currentThread);
+											}
+											if (NULL != methodObj) {
+												initImpl(currentThread, memberName, methodObj);
```
`memberName` would have been lost between being fetched and used here.

fengxue-IS commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1536 @@
+ * Given a MemberName object, try to set the uninitialized fields from existing values.
+ *
+ * Throws NullPointerException if MemberName object is null.
+ * Throws IllegalArgumentException if MemberName doesn't contain required data to expand.
```
updated comment

fengxue-IS commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1531 @@
+	while (J9ROMCLASS_IS_ARRAY(myClass->romClass)) {
+		J9Class * componentClass = (J9Class *)(((J9ArrayClass*)myClass)->componentType);
+		if (J9ROMCLASS_IS_PRIMITIVE_TYPE(componentClass->romClass)) {
+			break;
```
a primitive array class have the correct signature including the component class type

fengxue-IS commented:
Fix all issues in review comments, local build passed
@DanHeidinga can you please take a look?

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+										j9object_t memberName = J9JAVAARRAYOFOBJECT_LOAD(currentThread, resultsArray, index);
+										if (NULL == memberName) {
+											result = -99;
+											goto done;
```
Why `-99`?  Is there a meaning to this value?

fengxue-IS replied:
This is the value used by OJDK impl for this particular error case

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+				j9object_t membernameObject = J9_JNI_UNWRAP_REFERENCE(self);
+				jint flags = J9VMJAVALANGINVOKEMEMBERNAME_FLAGS(currentThread, membernameObject);
+				jlong vmindex = (jlong)(UDATA)J9OBJECT_U64_LOAD(currentThread, membernameObject, vm->vmindexOffset);
+				j9object_t target;
```
```suggestion
				j9object_t target = null;
```

fengxue-IS replied:
fixed

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+				}
+
+				J9VMJAVALANGLONG_SET_VALUE(currentThread, box, vmindex);
+				J9JAVAARRAYOFOBJECT_STORE(currentThread, arrayObject, 0, box);
+				J9JAVAARRAYOFOBJECT_STORE(currentThread, arrayObject, 0, target);
+
```
Why are both of these writing to index 0?

fengxue-IS replied:
fixed

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+ * static native void setCallSiteTargetVolatile(CallSite site, MethodHandle target);
+ */
+void JNICALL
+Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetVolatile(JNIEnv *env, jclass clazz, jobject callsite, jobject target)
```
This API and `Java_java_lang_invoke_MethodHandleNatives_setCallSiteTargetNormal` are identical except for a boolean.  Can you introduce a helper and pass the boolean to it?

fengxue-IS replied:
added common helper

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+		} else if (((start < -4) || (start > end) || (pos < 0)) || ((jint)J9INDEXABLEOBJECT_SIZE(currentThread, bufferArray) <= pos) || ((jint)J9INDEXABLEOBJECT_SIZE(currentThread, bufferArray) <= (pos + end - start))) {
+			vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR, NULL);
+		} else {
+			/* U_16 bsmArgCount = (U_16)J9JAVAARRAYOFINT_LOAD(currentThread, indexInfoArray, 0); */
```
Why the commented code?

fengxue-IS replied:
This should be identical to the argCount value from bsmData, I double checked against OJDK behavior and added a check that user value should be equal to jvm value.

fengxue-IS commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+ *
+ * Note: caller must have vmaccess before invoking this helper
+ */
+void
```
fixed

fengxue-IS commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+				if (J9ROMCLASS_IS_PRIMITIVE_TYPE(componentClass->romClass)) {
+					break;
+				}
+				numDims++;
```
fixed

fengxue-IS commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+			sigLength++; /* for null-termination */
+			sig = (char *)j9mem_allocate_memory(sigLength, OMRMEM_CATEGORY_VM);
+			if (NULL != sig) {
+				U_32 i;
```
fixed

fengxue-IS commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1529 @@
+	return sig;
+}
+
+char *
```
fixed

fengxue-IS commented:
Addressed review comments, local compile/basic testing passed.
@DanHeidinga do you see any other issues I should fix?

DanHeidinga commented on runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp:
```diff
@@ -0,0 +1,1527 @@
+		j9object_t callsiteObject = J9_JNI_UNWRAP_REFERENCE(callsite);
+		j9object_t targetObject = J9_JNI_UNWRAP_REFERENCE(target);
+
+		UDATA offset = (UDATA)vmFuncs->instanceFieldOffset(currentThread, J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, callsiteObject), (U_8*)"target", strlen("target"), (U_8*)"Ljava/lang/invoke/MethodHandle;", strlen("Ljava/lang/invoke/MethodHandle;"), NULL, NULL, 0);
+		MM_ObjectAccessBarrierAPI objectAccessBarrier = MM_ObjectAccessBarrierAPI(currentThread);
```
This seems like a good candidate to put these in the VMConstantpool

DanHeidinga commented:
I don't see any holes where objects will be lost now so I'm OK with proceeding with this change.

Having said that, I'm not 100% onboard with this approach.  The heavy string allocation and copying should be looked at again to see if there's a better way to solve this.  Given the other up-calls to the VM, I'd be tempted to implement this in Java and cache the signature per MT to make this cleaner.

All that can wait for a later PR though.

DanHeidinga commented:
Jenkins test sanity xlinux jdk16

DanHeidinga commented:
Jenkins test sanity xlinux jdk8,jdk11

fengxue-IS commented:
> Having said that, I'm not 100% onboard with this approach. The heavy string allocation and copying should be looked at again to see if there's a better way to solve this. Given the other up-calls to the VM, I'd be tempted to implement this in Java and cache the signature per MT to make this cleaner.
> 

Thanks for the suggestions Dan, I have added this to the TODO list (#11443) to be investigated later
